internal/application.go

```diff
@@ -0,0 +1,326 @@
+package internal
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"net/netip"
+	"strings"
+	"time"
+
+	"github.com/corazawaf/coraza/v3"
+	"github.com/corazawaf/coraza/v3/types"
+	"github.com/dropmorepackets/haproxy-go/pkg/encoding"
+	"github.com/rs/zerolog"
+	"istio.io/istio/pkg/cache"
+)
+
+type Application struct {
+	waf    coraza.WAF
+	logger *zerolog.Logger
+	cache  cache.ExpiringCache
+
+	ResponseCheck    bool
+	TransactionTTLMs time.Duration
+}
+
+type applicationRequest struct {
+	ID      string
+	SrcIp   netip.Addr
+	SrcPort int64
+	DstIp   netip.Addr
+	DstPort int64
+	Method  string
+	Path    []byte
+	Query   []byte
+	Version string
+	Headers []byte
+	Body    []byte
+}
+
+func (a *Application) HandleRequest(ctx context.Context, message *encoding.Message) error {
+	k := encoding.AcquireKVEntry()
+	defer encoding.ReleaseKVEntry(k)
+
+	var req applicationRequest
+	for message.KV.Next(k) {
+		switch name := string(k.NameBytes()); name {
+		case "id":
+			req.ID = string(k.ValueBytes())
+		case "src-ip":
+			req.SrcIp = k.ValueAddr()
+		case "src-port":
+			req.SrcPort = k.ValueInt()
+		case "dst-ip":
+			req.DstIp = k.ValueAddr()
+		case "dst-port":
+			req.DstPort = k.ValueInt()
+		case "method":
+			req.Method = string(k.ValueBytes())
+		case "path":
+			// make a copy of the pointer and add a defer in case there is another entry
+			currK := k
+			defer encoding.ReleaseKVEntry(currK)
+
+			req.Path = currK.ValueBytes()
+
+			// acquire a new kv entry to continue reading other message values.
+			k = encoding.AcquireKVEntry()
+		case "query":
+			// make a copy of the pointer and add a defer in case there is another entry
+			currK := k
+			defer encoding.ReleaseKVEntry(currK)
+
+			req.Query = currK.ValueBytes()
+			// acquire a new kv entry to continue reading other message values.
+			k = encoding.AcquireKVEntry()
+		case "version":
+			req.Version = string(k.ValueBytes())
+		case "headers":
+			// make a copy of the pointer and add a defer in case there is another entry
+			currK := k
+			defer encoding.ReleaseKVEntry(currK)
+
+			req.Headers = currK.ValueBytes()
+			// acquire a new kv entry to continue reading other message values.
+			k = encoding.AcquireKVEntry()
+		case "body":
+			// make a copy of the pointer and add a defer in case there is another entry
+			currK := k
+			defer encoding.ReleaseKVEntry(currK)
+
+			req.Body = currK.ValueBytes()
+			// acquire a new kv entry to continue reading other message values.
+			k = encoding.AcquireKVEntry()
+		default:
+			a.logger.Debug().Str("name", name).Msg("unknown kv entry")
+		}
+	}
+
+	if req.ID == "" {
+		return fmt.Errorf("request id is empty")
+	}
+
+	tx := a.waf.NewTransactionWithID(req.ID)
```

Comment:
I am usually reluctant about this, is req.ID something unique per instance e.g. in a distributed environment all hops in a request will reuse the X-Request-Id, not sure if this is the same thing or is it a unique value generated on every request to haproxy.

Reply:
This is the same behaviour as current [main](https://github.com/corazawaf/coraza-spoa/blob/main/internal/spoa.go#L279). The request ID is set via [unique-id-format](https://docs.haproxy.org/dev/configuration.html#4-unique-id-format) in HAProxy [here (e2e)](https://github.com/corazawaf/coraza-spoa/blob/rewrite/internal/e2e_test.go#L79) and [here (examples)](https://github.com/corazawaf/coraza-spoa/blob/rewrite/examples/haproxy.cfg#L19).

Reply:
Do you know if haproxy propagates it?

Reply:
Just to be on the same page, the Header will be injected into the request and send to the backend. I dont know what you mean with distributed environments

Reply:
Based on an idea by @mxey I moved the id generation to the request handler, set it as spop transaction value and pass it back inside the response handler. This removes the requirement of `unique-id-format`